x-pack/plugins/security/server/routes/api/v1/__tests__/authenticate.js

```diff
@@ -142,14 +123,12 @@ describe('Authentication routes', () => {
         const checkPrivilegesStub = sinon.stub().returns({ result: CHECK_PRIVILEGES_RESULT.AUTHORIZED });
         checkPrivilegesWithRequestStub.returns(checkPrivilegesStub);
 
-        await loginRoute.handler(request, replyStub);
+        await loginRoute.handler(request, hStub);
 
         sinon.assert.calledWithExactly(checkPrivilegesWithRequestStub, request);
         sinon.assert.calledWithExactly(checkPrivilegesStub, [serverStub.plugins.security.authorization.actions.login]);
         sinon.assert.neverCalledWith(serverStub.log, ['warning', 'deprecated', 'security'], getDeprecationMessage(user.username));
-        sinon.assert.notCalled(replyStub);
-        sinon.assert.calledOnce(replyStub.continue);
-        sinon.assert.calledWithExactly(replyStub.continue, { credentials: user });
+        sinon.assert.calledOnce(hStub.response);
```

Comment:
Is there a way to not have to stub out the `h` response toolkit, and instead do the asserts only based on the response value of the handler? If I understand correctly, the return values from the handlers should contain all of the information that we need if we could pass an actual implementation of `h` to the handlers. Not sure how Hapi 17 exposes this for tests like this.

I don't have any real issue with the way that you've done this here, just asking for future tests.

Reply:
In a test like this where we're using a server stub rather than an actual Hapi.Server instance, I couldn't find an _officially supported_ way to create a fake toolkit for the purpose of testing. Now we could use a real server instance and use `await server.inject` and then inspect it's return value.

I think testing like that, while it may be slower, does allow our tests to only verify the output rather than the implementation details. It would make refactors/package upgrades like this require many fewer test changes.

Reply:
>  Now we could use a real server instance and use await server.inject and then inspect it's return value.

> I think testing like that, while it may be slower, does allow our tests to only verify the output rather than the implementation details. It would make refactors/package upgrades like this require many fewer test changes.

Just wanted to note that we should not rely on `server.inject` for two main reasons:

1. In master all requests go through new platform/core, so bypassing core in tests with `kbnServer.server.inject` won't test real setup
2. We should _try_ to abstract away from Hapi specific APIs as much as possible since we would like to be able to switch to another HTTP framework in the future and tests should help us here, not slow us down

For that purpose I'm changing [`test_utils/kbn_server`](https://github.com/elastic/kibana/pull/22190/files#diff-89a18ea56e901a28a29e99fe9a697823) so that we can use `supertest` instead of `HapiServer.inject` [like that](https://github.com/elastic/kibana/pull/22190/files#diff-28b96eccb002d27f70dfeb7907bca289R45) that will work with any wrapper around native `http.Server` (ExpressJS, Hapi, Koa etc.). It still uses some hackery inside to access core's private properties, but it's hidden from consumers and we'll improve that at some point.

Reply:
Ah I like this a lot. Sounds like a great step forward. Hadn't thought about it, but I agree getting away from framework-specifics is better. Sounds like that in the time being, leaving this particular test as-is until we can use `supertest` is probably best.